front/components/agent_builder/capabilities/mcp/utils/dialogUtils.ts

```diff
@@ -0,0 +1,284 @@
+      },
+    };
+  }
+
+  return {};
+}
+
+export interface SaveConfigurationOptions {
+  mode: DialogMode | null;
+  configuredAction: AgentBuilderAction;
+  mcpServerView: MCPServerViewType;
+  onActionUpdate?: (action: AgentBuilderAction, index: number) => void;
+  onModeChange: (mode: DialogMode | null) => void;
+  setSelectedToolsInDialog: React.Dispatch<
+    React.SetStateAction<SelectedTool[]>
+  >;
+  setIsOpen: (open: boolean) => void;
+  sendNotification: (notification: {
+    title: string;
+    description: string;
+    type: "success" | "error";
+  }) => void;
+}
+
+export function handleConfigurationSave({
+  mode,
+  configuredAction,
+  mcpServerView,
+  onActionUpdate,
+  onModeChange,
+  setSelectedToolsInDialog,
+  setIsOpen,
+  sendNotification,
+}: SaveConfigurationOptions): void {
+  if (mode?.type === "edit" && onActionUpdate) {
+    // Edit mode: save the updated action and close dialog
+    onActionUpdate(configuredAction, mode.index);
+    setIsOpen(false);
+    onModeChange(null);
+
+    sendNotification({
+      title: "Tool updated successfully",
+      description: `${getMcpServerViewDisplayName(mcpServerView)} configuration has been updated.`,
+      type: "success",
+    });
+    return;
+  }
+
+  if (mode?.type === "configure") {
+    // Configure mode: add the configured tool to selected tools and go back to selection
+    setSelectedToolsInDialog((prev) => {
+      const existingToolIndex = prev.findIndex(
+        (tool) =>
+          tool.type === "MCP" &&
+          tool.configuredAction?.name === configuredAction.name
+      );
+
+      if (existingToolIndex >= 0) {
+        // Update existing tool with configuration
+        const updated = [...prev];
+        updated[existingToolIndex] = {
+          type: "MCP",
+          view: mcpServerView,
+          configuredAction,
+        };
+        return updated;
+      } else {
+        // Add new configured tool
+        return [
+          ...prev,
+          {
+            type: "MCP",
+            view: mcpServerView,
+            configuredAction,
+          },
+        ];
+      }
+    });
+
+    // Go back to add mode and tool selection
+    onModeChange({ type: "add" });
+    return;
+  }
+
+  // Default case (add mode or other): add to selected tools and navigate back
+  setSelectedToolsInDialog((prev) => {
+    const existingToolIndex = prev.findIndex(
+      (tool) =>
+        tool.type === "MCP" &&
+        tool.configuredAction?.name === configuredAction.name
+    );
+
+    if (existingToolIndex >= 0) {
+      // Update existing tool with configuration
+      const updated = [...prev];
+      updated[existingToolIndex] = {
+        type: "MCP",
+        view: mcpServerView,
+        configuredAction,
+      };
+      return updated;
+    } else {
+      // Add new configured tool
+      return [
+        ...prev,
+        {
+          type: "MCP",
+          view: mcpServerView,
+          configuredAction,
+        },
+      ];
+    }
+  });
```

Comment:
Seems that the code inside `setSelectedToolsInDialog` are exactly the same?
We could first set the tools, and if `type === configuration` switch the mode?

```suggestion
setSelectedToolsInDialog((prev) => {
    const existingToolIndex = prev.findIndex(
      (tool) =>
        tool.type === "MCP" &&
        tool.configuredAction?.name === configuredAction.name
    );

    if (existingToolIndex >= 0) {
      // Update existing tool with configuration
      const updated = [...prev];
      updated[existingToolIndex] = {
        type: "MCP",
        view: mcpServerView,
        configuredAction,
      };
      return updated;
    } else {
      // Add new configured tool
      return [
        ...prev,
        {
          type: "MCP",
          view: mcpServerView,
          configuredAction,
        },
      ];
    }
  });
  
  if (mode?.type === "configure") {
    // Go back to add mode and tool selection
    onModeChange({ type: "add" });
  }
  
```

Reply:
Indeed! Good catch thanks